src/main/java/javax/jmdns/impl/ServiceInfoImpl.java

```diff
@@ -418,15 +418,35 @@ public String getHostAddress() {
      */
     @Override
     public String[] getHostAddresses() {
-        Inet4Address[] ip4Aaddresses = this.getInet4Addresses();
-        Inet6Address[] ip6Aaddresses = this.getInet6Addresses();
-        String[] names = new String[ip4Aaddresses.length + ip6Aaddresses.length];
-        for (int i = 0; i < ip4Aaddresses.length; i++) {
-            names[i] = ip4Aaddresses[i].getHostAddress();
+        Inet4Address[] ip4Addresses = this.getInet4Addresses();
+        Inet6Address[] ip6Addresses = this.getInet6Addresses();
+
+        // Default to empty arrays if null
+        if (ip4Addresses == null) {
+            ip4Addresses = new Inet4Address[0];
+        }
+        if (ip6Addresses == null) {
+            ip6Addresses = new Inet6Address[0];
         }
-        for (int i = 0; i < ip6Aaddresses.length; i++) {
-            names[i + ip4Aaddresses.length] = "[" + ip6Aaddresses[i].getHostAddress() + "]";
+
+        String[] names = new String[ip4Addresses.length + ip6Addresses.length];
```

Comment:
This is what I referred to with "which will require changes to how the array is constructed and filled" - your array size is now still the same, but you might have empty entries for certain indexes, which is not good.
Imho, the array should only contain valid IP addresses and have no gaps inbetween, would you agree?

Reply:
agreed.  I missed that point.

Reply:
OK i took a different approach where it has to count them up front to get the proper size and then add the data into the array

Reply:
Tbh, iterating multiple times over the same lists doesn't look very elegant to me.
How about simply using `names = ArrayList<>()` and converting it to an array before returning the result (`names.toArray()`)?

Reply:
Good idea how does it look now?

Reply:
That looks lovely! ❤️